Change sanity check used in known_hosts module

ptMcGit opened the pull request:
##### SUMMARY

Previously `sanity_check()` in the known_hosts module used
`ssh-keygen -F -f keyfile` to validate the keyfile.

This command treats the keyfile as a known hosts file
and checks for the host. However, this fails with valid keys
in the OpenSSH .pub format.

`sanity_check()` now uses `ssh-keygen -l -f keyfile` to validate the
keyfile. When `ssh-keygen` fails to detect a valid key it returns a
non-zero exit status which is used to trigger module.fail_json().

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

modules/system/known_hosts.py

##### ANSIBLE VERSION

  config file = /Users/user/.ansible.cfg
  configured module search path = [u'/Users/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.4.1.0/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Nov 11 2017, 11:37:54) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)]

##### ADDITIONAL INFORMATION

Recreate:

1. Create a test key...
2. Use with `known_hosts` module:

``` YAML
- hosts: all
  tasks:
    - name: Add host key.
      known_hosts:
      name: "myhost"
      key: "{{ lookup('file', 'mytestkey.pub') }}"
      state: present
```

Output:

```
TASK [Add host key.] ***************************************************************************************************************************************
fatal: [practice]: FAILED! => {"changed": false, "cmd": "/usr/bin/ssh-keygen -F myhost -f /tmp/xxxxxxxxx", "failed": true, "msg": "", "rc": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
```

Adding 'myhost' to beginning of key file alleviates the problem, but is not ideal:

```
TASK [Add host key.] ***************************************************************************************************************************************
changed: [practice]
```

After applying the fix, it works, and uses `ssh-keygen` to flag bogus keys:

```
TASK [Add host key.] ***************************************************************************************************************************************
fatal: [practice]: FAILED! => {"changed": false, "failed": true, "msg": "ssh-rsa asdfasdfadjlkf\n is not an acceptable public key file."}
```

Unit test results:

```
=================================================================== test session starts ====================================================================
platform linux2 -- Python 2.7.12, pytest-3.3.1, py-1.5.2, pluggy-0.6.0
rootdir: /home/ubuntu/ansible, inifile:
plugins: xdist-1.20.1, mock-1.6.3, forked-0.2, f5-sdk-3.0.6
collected 6 items                                                                                                                                          

test/units/modules/system/test_known_hosts.py ......                                                                                                 [100%]

------------------------------------- generated xml file: /home/ubuntu/ansible/test/results/junit/python2.7-units.xml --------------------------------------
================================================================= 6 passed in 0.20 seconds =================================================================
_________________________________________________________________________ summary __________________________________________________________________________
  py27: commands succeeded
  congratulations :)
```

ptMcGit committed:
Change sanity check used in known_hosts module

Previously `sanity_check()` in the known_hosts module used
`ssh-keygen -F -f keyfile` to validate the keyfile.

This command treats the keyfile as a known hosts file
and checks for the host. However, this fails with valid keys
in the OpenSSH .pub format.

`sanity_check()` now uses `ssh-keygen -l -f keyfile` to validate the
keyfile. When `ssh-keygen` fails to detect a valid key it returns a
non-zero exit status which is used to trigger module.fail_json().

robinro commented:
This PR removes relevant functionality, since it removes the check for a matching hostname in the given key, which prevents incorrect keys inserted by accident.

@ptMcGit Can you add the hostname explicitly instead? Does the following work for you?
```
- hosts: all
  tasks:
    - name: Add host key.
      known_hosts:
      name: "myhost"
      key: "myhost {{ lookup('file', 'mytestkey.pub') }}"
      state: present
```

ptMcGit commented:
@robinro, thanks for clarifying this. However, I don't see how the following guarantees that the key is for `myhost` as the following works as well:

``` YAML
---
- hosts: all
  tasks:
    - name: Add host key.
      known_hosts:
        name: "super"
        key: "super {{ lookup('file', 'mytestkey.pub') }}"
        state: present
```

What do you think about checking for the host in the comment field to achieve this?

Something in the spirit of:

``` Python
key = re.sub(r'\s+', ' ',key)                                                # remove whitespace                                                                                        
ssh_host_in_comment = key.split(' ')[-2].split('@')[-1]    # get the host         
                                                                                                                                                                      
if ssh_host_in_comment != name:                                                                                                                                       
    module.fail_json(msg="The name provided was not indicated in the supplied key")                                                                                                    
```
